src/tests/test_method_middleware.py

```diff
@@ -0,0 +1,261 @@
+from typing import Callable
+
+import pytest
+
+from polydatum import DataAccessLayer, Service
+from polydatum.dal import DataManager
+from polydatum.middleware import DalCommandRequest, DalCommand, dal_method_resolver_middleware, \
+    handle_dal_method, dal_resolver, DalMethodError
+from polydatum.errors import InvalidMiddleware
+
+
+class FooMiddleware:
+    # Does not implement __call__ and hence not
+    # a Callable
+    pass
+
+
+@pytest.mark.parametrize('middleware', [
+    (FooMiddleware, ),
+    (FooMiddleware(), ),
+    (True, ),
+    ('random-string',),
+    ([],),
+    (None,),
+])
+def test_dal_middleware_requires_callable(middleware):
+    """
+    Verify that you can specify method middleware as either an instance
+    or a class (as a convenience).
+    """
+    with pytest.raises(InvalidMiddleware):
+        DataAccessLayer(data_manager=None, middleware=[middleware], default_middleware=None)
+
+
+def test_dal_method_middleware():
+    """
+    Verify that when calling a method on a service off the DAL,
+    the method middlewares run, in the correct order and run
+    before the method itself is called.
+
+
+    We deliberately do not use the DalMethodResolverMiddleware here,
+    because we are testing middleware call chains, not functionality
+    of that class.
+    """
+    def test_method():
+        return ['default-handler-call']
+
+    class OuterMiddleware:
+        def __call__(self, request, handler):
+            result = ['outer-middleware-ingress']
+            result.extend(handler(request))
+            result.append('outer-middleware-egress')
+            return result
+
+    class SecondMiddleware:
+        def __call__(self, request, handler):
+            result = ['second-middleware-ingress']
+            result.extend(handler(request))
+            result.append('second-middleware-egress')
+            return result
+
+    # Verify custom default middleware
+    class DefaultMiddleware:
+        def __call__(self, request, handler):
+            result = ['default-middleware-ingress']
+            # We need to get past the DalMethodResolverMiddleware
+            # which sets the `dal_method` attribute to
+            # the method that has to be called
+            request.dal_method = test_method
+            result.extend(handler(request))
+            result.append('default-middleware-egress')
+            return result
+
+    dm = DataManager()
+
+    dal = DataAccessLayer(data_manager=dm, middleware=[OuterMiddleware(), SecondMiddleware()],
+        default_middleware=(DefaultMiddleware(),))
+
+    with dm.context():
+        result = dal.fake.service.method()
+
```

Comment:
After this, adding this would verify get item access worked with middleware:

```
    with dm.context():
        result = dal.fake.service.method()
        assert dal["fake.service.method"]() == result, "Get item access should run the same middleware"
```